# NATURAL LANGUAGE PROCESSING (NLP)

_Natural Language Processing (NLP) helps computers understand, interpret and produce human language. It studies language as data and develops a model that can analyse linguistic structure, meaning and context in both written and spoken communication._

#### TEXT PREPROCESSING:-

In [1]:
text = "Hii!!! I luv AI 😊. Visit https://example.com <p>NOW</p> u r gr8!!!"

**1. Lower Casing:**
_Because in Machine 'B' and 'b' treated as different to simplify it we convert all text in lower case._

In [2]:
lower_text = text.lower()
print(lower_text)

hii!!! i luv ai 😊. visit https://example.com <p>now</p> u r gr8!!!


**2. Remove HTML Tags (Using RE):**
_This function removes HTML/XML tags from the given text by using Regular Expression. We can change the **RE: <.*?>** as per the data.


In [3]:
import re

no_tags = re.sub(r'<.*?>', '', lower_text)
print(no_tags)

hii!!! i luv ai 😊. visit https://example.com now u r gr8!!!


**3. Remove URLs (Using RE):** Use to remove URLs

In [4]:
no_urls = re.sub(r'http\S+|www\S+', '', no_tags)
print(no_urls)

hii!!! i luv ai 😊. visit  now u r gr8!!!


**4. Remove Punctuation:** Use to remove Punctuation

In [5]:
import string

no_punct = no_urls.translate(str.maketrans('', '', string.punctuation))
print(no_punct)

hii i luv ai 😊 visit  now u r gr8


**5. Chat Word Treatment:** Use to deal with Genz words.

In [6]:
chat_dict = {
    "luv": "love",
    "u": "you",
    "r": "are",
    "gr8": "great"
}

words = no_punct.split()
normalized = [chat_dict.get(word, word) for word in words]

chat_clean = " ".join(normalized)
print(chat_clean)

hii i love ai 😊 visit now you are great


**6. Spelling Correction:**

TextBlob is a simple Python library for processing and analyzing text data. It builds on NLTK and Pattern, providing an easy to use interface for tasks like tokenization, part of speech tagging, sentiment analysis, translation and noun phrase extraction.

**TextBlob.correct()** is a method in the TextBlob library that attempts to automatically fix spelling mistakes in a given sentence. It analyzes each word and predicts the most likely correct spelling. However, this method works well only for very simple spelling errors and may fail or give unexpected results on complex sentences.

In [7]:
from textblob import TextBlob

corrected = str(TextBlob(chat_clean).correct())
print(corrected)

his i love ai 😊 visit now you are great


**7. Remove Stopwords:**

Natural language processing tasks often involve filtering out commonly occurring words that provide no or very little semantic value to text analysis. These words are known as stopwords include articles, prepositions and pronouns like "the", "and", "is" and "in". While they seem insignificant, proper stopword handling can dramatically impact the performance and accuracy of NLP applications.

In [8]:
from nltk.corpus import stopwords

stop_words = set(stopwords.words('english'))

words = corrected.split()
filtered = [w for w in words if w not in stop_words]

stop_clean = " ".join(filtered)
print(stop_clean)

love ai 😊 visit great


**8. Emoji Handling (Convert to Text):** Converting it into text

In [12]:
import emoji

final_text = emoji.demojize(stop_clean)
print(final_text)

love ai :smiling_face_with_smiling_eyes: visit great


### Stemming:

Stemming is an important text-processing technique that reduces words to their base or root form by removing prefixes and suffixes.


##### 1. Porter's Stemmer:

Porter's Stemmer is one of the most popular and widely used stemming algorithms. This stemmer works by applying a series of rules to remove common suffixes from English words. It is well-known for its simplicity, speed and reliability.

In [13]:
from nltk.stem import PorterStemmer

porter_stemmer = PorterStemmer()

words = ["running", "jumps", "happily", "running", "happily"]

stemmed_words = [porter_stemmer.stem(word) for word in words]

print("Original words:", words)
print("Stemmed words:", stemmed_words)

Original words: ['running', 'jumps', 'happily', 'running', 'happily']
Stemmed words: ['run', 'jump', 'happili', 'run', 'happili']



##### 2. Snowball Stemmer:

The Snowball Stemmer is an enhanced version of the Porter Stemmer which was introduced by Martin Porter as well. It is referred to as Porter2 and is faster and more aggressive than its predecessor. One of the key advantages of this is that it supports **multiple languages, making it a multilingual stemmer.**

In [14]:
from nltk.stem import SnowballStemmer

stemmer = SnowballStemmer(language='english')

words_to_stem = ['running', 'jumped', 'happily', 'quickly', 'foxes']

stemmed_words = [stemmer.stem(word) for word in words_to_stem]

print("Original words:", words_to_stem)
print("Stemmed words:", stemmed_words)

Original words: ['running', 'jumped', 'happily', 'quickly', 'foxes']
Stemmed words: ['run', 'jump', 'happili', 'quick', 'fox']



##### 3. Lancaster Stemmer:

The Lancaster Stemmer is known for being more aggressive and faster than other stemmers. However, it’s also more destructive and may lead to excessively shortened stems. It uses a set of external rules that are applied in an iterative manner.Less efficient than Snowball in larger datasets.


In [15]:
from nltk.stem import LancasterStemmer

stemmer = LancasterStemmer()

words_to_stem = ['running', 'jumped', 'happily', 'quickly', 'foxes']

stemmed_words = [stemmer.stem(word) for word in words_to_stem]

print("Original words:", words_to_stem)
print("Stemmed words:", stemmed_words)

Original words: ['running', 'jumped', 'happily', 'quickly', 'foxes']
Stemmed words: ['run', 'jump', 'happy', 'quick', 'fox']



##### 4. Regexp Stemmer:

The Regexp Stemmer or Regular Expression Stemmer is a flexible stemming algorithm that allows users to define custom rules using regular expressions (regex). This stemmer can be helpful for very specific tasks where predefined rules are necessary for stemming.

In [16]:
from nltk.stem import RegexpStemmer

custom_rule = r'ing$' #Custom Rule defined i.e remove 'ing'
regexp_stemmer = RegexpStemmer(custom_rule)

word = 'running'
stemmed_word = regexp_stemmer.stem(word)

print(f'Original Word: {word}')
print(f'Stemmed Word: {stemmed_word}')

Original Word: running
Stemmed Word: runn


### Lemmatization:

Lemmatization is an important text pre-processing technique in Natural Language Processing (NLP) that reduces words to their base form known as a "lemma." For example, the lemma of "running" is "run" and "better" becomes "good."

Unlike stemming which simply removes prefixes or suffixes, it considers the word's meaning and part of **speech (POS)** and ensures that the base form is a valid word. This makes lemmatization more accurate as it avoids generating non-dictionary words.

It is used for:

* **Improves accuracy:** It ensures words with similar meanings like "running" and "ran" are treated as the same.

* **Reduced Data Redundancy:** By reducing words to their base forms, it reduces redundancy in the dataset. This leads to smaller datasets which makes it easier to handle and process large amounts of text for analysis or training machine learning models.

* **Better NLP Model Performance:** By treating all similar word as same, it improves the performance of NLP models by making text more consistent. For example, treating "running," "ran" and "runs" as the same word improves the model's understanding of context and meaning.

In [17]:
import nltk
from nltk.stem import WordNetLemmatizer

# Download required data
nltk.download('wordnet')
nltk.download('omw-1.4')

# Initialize lemmatizer
lemmatizer = WordNetLemmatizer()

# Example words
words = ["running", "studies", "better", "cars"]

# Apply lemmatization
lemmatized = [lemmatizer.lemmatize(word) for word in words]

print("Original:", words)
print("Lemmatized:", lemmatized)

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\kushp\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\kushp\AppData\Roaming\nltk_data...


Original: ['running', 'studies', 'better', 'cars']
Lemmatized: ['running', 'study', 'better', 'car']


#### VECTORIZATION:

* Vectorization in NLP is the process of converting unstructured text data into numerical vectors (lists of numbers) that machine learning models can process.
* It is essential for feature extraction, enabling tasks like classification and semantic analysis.
* Key techniques include Count Vectorization (Bag of Words), TF-IDF, and dense word embeddings (Word2Vec, GloVe).


### 1. One-Hot Encoding:

One-Hot Encoding is a technique where each word is represented by a vector with a high bit corresponding to the word’s index in the vocabulary with all other bits set to zero.

##### Advantages of One-Hot Encoding:

* **Simplicity:** Easy to implement and understand.
* **Interpretability:** The vectors are easily interpretable and directly map to words.
* **Compatibility:** Works well with most machine learning algorithms.

##### Disadvantages of One-Hot Encoding:

* **High Dimensionality:** Results in large, sparse vectors for large vocabularies.
* **Loss of Semantic Information:** Does not capture word relationships or meaning.
* **Sparsity:** Creates sparse vectors with mostly zero values, leading to inefficiency.

In [1]:
from sklearn.preprocessing import OneHotEncoder
import numpy as np
import string

documents = [
    "The cat sat on the mat.",
    "The dog sat on the log.",
    "Cats and dogs are pets."
]

words = [word.lower().strip(string.punctuation)
         for doc in documents for word in doc.split()]
vocabulary = sorted(set(words))

encoder = OneHotEncoder(sparse_output=False)  ## sparse_output=False → gives output as normal array (not sparse matrix)
one_hot_vectors = encoder.fit_transform(np.array(vocabulary).reshape(-1, 1))

word_to_onehot = {vocabulary[i]: one_hot_vectors[i] #Create mapping of word → one-hot vector
                  for i in range(len(vocabulary))}

for word, vector in word_to_onehot.items():
    print(f"Word: {word}, One-Hot Encoding: {vector}")

Word: and, One-Hot Encoding: [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
Word: are, One-Hot Encoding: [0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
Word: cat, One-Hot Encoding: [0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
Word: cats, One-Hot Encoding: [0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
Word: dog, One-Hot Encoding: [0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0.]
Word: dogs, One-Hot Encoding: [0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0.]
Word: log, One-Hot Encoding: [0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]
Word: mat, One-Hot Encoding: [0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]
Word: on, One-Hot Encoding: [0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0.]
Word: pets, One-Hot Encoding: [0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]
Word: sat, One-Hot Encoding: [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0.]
Word: the, One-Hot Encoding: [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]



### 2. Bag of Words (BoW):

Bag of Words (BoW) converts text into a vector representing the frequency of words, disregarding grammar and word order. It **counts the occurrences of each word in a document** and generates a vector based on these counts.

##### Advantages of Bag of Words (BoW):

* Simple and easy to implement.
* Provides a clear and interpretable representation of text.

##### Disadvantages of Bag of Words (BoW):

* Ignores the order and context of words.
* Results in high-dimensional and sparse matrices.
* Fails to capture semantic meaning and relationships between words.

In [2]:
from sklearn.feature_extraction.text import CountVectorizer

documents = [
    "The cat sat on the mat.",
    "The dog sat on the log.",
    "Cats and dogs are pets."
]

vectorizer = CountVectorizer()
X = vectorizer.fit_transform(documents)

print(X.toarray())
print(vectorizer.get_feature_names_out())

[[0 0 1 0 0 0 0 1 1 0 1 2]
 [0 0 0 0 1 0 1 0 1 0 1 2]
 [1 1 0 1 0 1 0 0 0 1 0 0]]
['and' 'are' 'cat' 'cats' 'dog' 'dogs' 'log' 'mat' 'on' 'pets' 'sat' 'the']



### 3. Term Frequency-Inverse Document Frequency (TF-IDF):

TF-IDF is an extension of BoW that weighs the frequency of words by their importance across documents.

##### 1. Term Frequency (TF):
 Measures the frequency of a word in a document.

**TF(t,d) = (Number of times t appear in document d) / (Total numbers of document in d)**



##### 2. Inverse Document Frequency (IDF):
Measures the importance of a word across the entire corpus.

**IDF(t) = log(Total number of Document/Number of Document contaning term t)**

##### Advantages of TF-IDF

* **Simple and Efficient:** Easy to compute with minimal computational resources.
* **Importance Weighting:** Highlights important words by reducing the influence of common words.
* **Effective for Information Retrieval:** Improves document relevance ranking in search engines.

##### Disadvantages of TF-IDF

* **Sparsity:** Produces high-dimensional sparse vectors, increasing memory and computational cost.
* **No Context Capture:** Does not account for word context or meaning.
* **Synonym Handling:** Treats synonyms as separate which may reduce model accuracy.

In [4]:
from sklearn.feature_extraction.text import TfidfVectorizer

documents = [
    "The cat sat on the mat.",
    "The dog sat on the log.",
    "Cats and dogs are pets."
]

tfidf_vectorizer = TfidfVectorizer() # It converts text → weighted numerical vectors
X_tfidf = tfidf_vectorizer.fit_transform(documents)

print(X_tfidf.toarray()) # Convert sparse matrix to dense array
print(tfidf_vectorizer.get_feature_names_out())

[[0.         0.         0.42755362 0.         0.         0.
  0.         0.42755362 0.32516555 0.         0.32516555 0.6503311 ]
 [0.         0.         0.         0.         0.42755362 0.
  0.42755362 0.         0.32516555 0.         0.32516555 0.6503311 ]
 [0.4472136  0.4472136  0.         0.4472136  0.         0.4472136
  0.         0.         0.         0.4472136  0.         0.        ]]
['and' 'are' 'cat' 'cats' 'dog' 'dogs' 'log' 'mat' 'on' 'pets' 'sat' 'the']


#### Cosine Similarity:
Cosine similarity measures the similarity between two non-zero vectors by calculating the cosine of the angle between them. It is widely used in machine learning and data analysis, especially in text analysis, document comparison, search queries, and recommendation systems.

* Similarity measure calculates the distance between data objects based on their feature dimensions in a dataset.
* A smaller distance indicates a higher similarity, while a larger distance indicates a lower similarity

**Cosine similarity works well even if:**

* Sentence length is different
* Word frequency varies

In [6]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# 🔹 Sample MeetMux User Bios
bios = [
    "Expert in Python and Machine Learning for social good.",
    "Professional Chef who loves outdoor Hiking and mountains.",
    "Machine Learning enthusiast and mountain hiker."
]

# 🔹 Step 1: Convert text → TF-IDF vectors
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(bios)

# 🔹 Step 2: Compute Cosine Similarity
similarity_matrix = cosine_similarity(tfidf_matrix)

# 🔹 Print similarity matrix
print("Cosine Similarity Matrix:\n", similarity_matrix)

Cosine Similarity Matrix:
 [[1.         0.04696865 0.2589084 ]
 [0.04696865 1.         0.06062112]
 [0.2589084  0.06062112 1.        ]]


#### Observations from Cosine Similarity Matrix:

* Diagonal values are 1, which means each user is 100% similar to themselves.
* The highest similarity (0.2589) is between User 0 and User 2.
* Indicates they share common interests (Machine Learning).
* User 1 has very low similarity with both users.
* (0.0469 with User 0, 0.0606 with User 2).
* Shows User 1 belongs to a different domain (Chef/Hiking).
* User 0 and User 2 are mutual best matches.
* Both have overlapping keywords → better compatibility.
* All similarity values are less than 0.3.
* Indicates dataset is small and diverse.
* TF-IDF successfully reduced importance of common words like “the”, “and”.
* Matching is based on shared important words, not sentence structure.


## Day 18: Interest Extraction from Text

**Polarity and Subjectivity:** Are core metrics in Natural Language Processing (NLP) sentiment analysis used to evaluate the tone and nature of text. Polarity measures sentiment intensity on a scale from negative to positive (e.g., -1.0 to +1.0), determining if a text is positive, negative, or neutral. Subjectivity measures how much of the text is personal opinion rather than factual information, often scaled from 0 (objective/factual) to 1 (subjective/opinionated).

In [7]:
from textblob import TextBlob

#Sample MeetMux User Bios / Reviews
bios = [
    "I absolutely loved the event, it was amazing and well organized!",
    "The event was okay, not too bad but could be better.",
    "I had a terrible experience, very poorly managed.",
    "Great vibes and friendly people, really enjoyed it!",
    "Not worth the time, very disappointing and boring."
]

#Analyze Sentiment
for i, bio in enumerate(bios):
    analysis = TextBlob(bio)

    polarity = analysis.sentiment.polarity
    subjectivity = analysis.sentiment.subjectivity

    print(f"\nBio {i+1}: {bio}")
    print(f"Polarity: {polarity}")
    print(f"Subjectivity: {subjectivity}")


Bio 1: I absolutely loved the event, it was amazing and well organized!
Polarity: 0.7250000000000001
Subjectivity: 0.8500000000000001

Bio 2: The event was okay, not too bad but could be better.
Polarity: 0.10000000000000005
Subjectivity: 0.5555555555555555

Bio 3: I had a terrible experience, very poorly managed.
Polarity: -0.76
Subjectivity: 0.89

Bio 4: Great vibes and friendly people, really enjoyed it!
Polarity: 0.6
Subjectivity: 0.65

Bio 5: Not worth the time, very disappointing and boring.
Polarity: -0.6433333333333334
Subjectivity: 0.6699999999999999



#### Reflection:
In a Markdown cell, answer: "How can MeetMux use Sentiment Analysis on
event reviews to automatically flag organizers who are providing a poor user
experience?"


MeetMux can use sentiment analysis on event reviews to automatically detect user satisfaction levels.

By analyzing the polarity of reviews:
- Negative sentiment (low polarity) can indicate poor user experience
- Positive sentiment reflects successful and engaging events

The platform can:
- Automatically flag organizers receiving consistently negative feedback
- Alert admins for quality checks
- Improve recommendation systems by promoting highly rated events
- Provide insights to organizers to improve future events

This helps ensure better user experience and builds trust within the platform.